File: backend/src/models/runSession.js
const mongoose = require('mongoose');

const messageSchema = new mongoose.Schema({
  sender: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  content: {
    type: String,
    required: true
  },
  timestamp: {
    type: Date,
    default: Date.now
  },
  type: {
    type: String,
    enum: ['text', 'system', 'achievement'],
    default: 'text'
  }
});

const checkpointSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true
  },
  description: String,
  location: {
    type: {
      type: String,
      enum: ['Point'],
      default: 'Point'
    },
    coordinates: {
      type: [Number],
      required: true
    }
  },
  radius: {
    type: Number,
    default: 20 // meters
  },
  order: {
    type: Number,
    required: true
  },
  participantProgress: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    reachedAt: Date
  }]
});

const locationSchema = new mongoose.Schema({
  type: {
    type: String,
    enum: ['Point'],
    default: 'Point'
  },
  coordinates: {
    type: [Number],  // [longitude, latitude]
    required: true
  },
  altitude: {
    type: Number,
    default: 0
  },
  speed: {
    type: Number,
    default: 0
  },
  timestamp: {
    type: Date,
    default: Date.now
  }
});

const runSessionSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  title: {
    type: String,
    required: true
  },
  description: String,
  startTime: {
    type: Date,
    default: Date.now
  },
  endTime: Date,
  status: {
    type: String,
    enum: ['planned', 'active', 'paused', 'completed', 'cancelled'],
    default: 'planned'
  },
  type: {
    type: String,
    enum: ['solo', 'group', 'challenge'],
    default: 'solo'
  },
  runStyle: {
    type: String,
    enum: ['free', 'checkpoint', 'race'],
    default: 'free'
  },
  participants: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    role: {
      type: String,
      enum: ['host', 'participant'],
      default: 'participant'
    },
    joinedAt: {
      type: Date,
      default: Date.now
    },
    lastLocation: locationSchema,
    isActive: {
      type: Boolean,
      default: true
    }
  }],
  maxParticipants: {
    type: Number,
    default: 1
  },
  chat: [messageSchema],
  checkpoints: [checkpointSchema],
  locationHistory: [locationSchema],
  paceHistory: {
    type: Map,
    of: [{
      pace: Number,
      timestamp: Date
    }]
  },
  currentDistance: {
    type: Map,
    of: Number
  },
  scheduledStart: {
    type: Date
  },
  weather: {
    temperature: Number,
    condition: String,
    windSpeed: Number,
    humidity: Number
  },
  metrics: [{
    distance: Number, // in meters
    duration: Number, // in seconds
    pace: Number, // in minutes per kilometer
    speed: Number, // in kilometers per hour
    calories: Number,
    elevation: Number, // in meters
    heartRate: Number, // beats per minute
    timestamp: {
      type: Date,
      default: Date.now
    }
  }],
  finalMetrics: {
    totalDistance: Number,
    totalDuration: Number,
    averagePace: Number,
    averageSpeed: Number,
    totalCalories: Number,
    elevationGain: Number,
    averageHeartRate: Number
  },
  route: {
    name: String,
    difficulty: {
      type: String,
      enum: ['easy', 'moderate', 'hard', 'extreme']
    },
    terrain: {
      type: String,
      enum: ['road', 'trail', 'track', 'mixed']
    }
  },
  photos: [{
    url: String,
    caption: String,
    location: {
      type: [Number], // [longitude, latitude]
    },
    timestamp: {
      type: Date,
      default: Date.now
    }
  }],
  comments: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    text: String,
    timestamp: {
      type: Date,
      default: Date.now
    }
  }],
  likes: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    timestamp: {
      type: Date,
      default: Date.now
    }
  }],
  privacy: {
    type: String,
    enum: ['public', 'friends', 'private'],
    default: 'public'
  },
  tags: [String]
}, {
  timestamps: true
});

// Index for geospatial queries on checkpoints
runSessionSchema.index({ 'checkpoints.location': '2dsphere' });

// Indexes for better query performance
runSessionSchema.index({ user: 1, startTime: -1 });
runSessionSchema.index({ status: 1 });
runSessionSchema.index({ 'participants.user': 1 });
runSessionSchema.index({ privacy: 1 });
runSessionSchema.index({ tags: 1 });

// Virtual for calculating current pace
runSessionSchema.virtual('currentPace').get(function() {
  if (this.metrics && this.metrics.length > 0) {
    return this.metrics[this.metrics.length - 1].pace;
  }
  return 0;
});

// Method to calculate session statistics
runSessionSchema.methods.calculateStatistics = function() {
  if (!this.metrics || this.metrics.length === 0) return null;

  const stats = {
    totalDistance: 0,
    totalDuration: 0,
    averagePace: 0,
    averageSpeed: 0,
    totalCalories: 0,
    elevationGain: 0,
    averageHeartRate: 0
  };

  let heartRateReadings = 0;

  this.metrics.forEach(metric => {
    stats.totalDistance += metric.distance || 0;
    stats.totalDuration += metric.duration || 0;
    stats.totalCalories += metric.calories || 0;
    if (metric.elevation > 0) stats.elevationGain += metric.elevation;
    if (metric.heartRate) {
      stats.averageHeartRate += metric.heartRate;
      heartRateReadings++;
    }
  });

  if (stats.totalDuration > 0) {
    stats.averageSpeed = (stats.totalDistance / 1000) / (stats.totalDuration / 3600);
    stats.averagePace = stats.totalDuration / (stats.totalDistance / 1000) / 60;
  }

  if (heartRateReadings > 0) {
    stats.averageHeartRate = stats.averageHeartRate / heartRateReadings;
  }

  return stats;
};

const RunSession = mongoose.model('RunSession', runSessionSchema);

module.exports = RunSession;

File: lib/features/run/data/models/run_session.dart
import 'package:json_annotation/json_annotation.dart';
import '../../../../core/services/user_service.dart';

part 'run_session.g.dart';

@JsonSerializable()
class RunSession {
  @JsonKey(name: '_id')
  final String? id;
  final String title;
  final String description;
  final User user;
  final List<Participant> participants;
  final String status;
  final String type;
  final String runStyle;
  final DateTime startTime;
  final DateTime? endTime;
  final DateTime scheduledStart;
  final bool isActive;
  final List<Checkpoint> checkpoints;
  final List<ChatMessage> chat;
  final int maxParticipants;
  final String privacy;
  final List<String> tags;
  final List<RunMetrics> metrics;
  final List<dynamic> photos;
  final List<dynamic> comments;
  final List<dynamic> likes;
  final DateTime createdAt;
  final DateTime updatedAt;

  bool get isParticipant => 
    user.id != null && 
    participants.any((p) => p.user.id == user.id);

  RunSession({
    this.id,
    required this.title,
    required this.description,
    required this.user,
    required this.participants,
    required this.status,
    required this.type,
    required this.runStyle,
    required this.startTime,
    this.endTime,
    required this.scheduledStart,
    this.isActive = false,
    required this.checkpoints,
    required this.chat,
    required this.maxParticipants,
    required this.privacy,
    required this.tags,
    required this.metrics,
    required this.photos,
    required this.comments,
    required this.likes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory RunSession.fromJson(Map<String, dynamic> json) =>
      _$RunSessionFromJson(json);

  Map<String, dynamic> toJson() => _$RunSessionToJson(this);
}

@JsonSerializable()
class User {
  @JsonKey(name: '_id')
  final String id;
  final String username;

  User({
    required this.id,
    required this.username,
  });

  factory User.fromJson(Map<String, dynamic> json) => _$UserFromJson(json);
  Map<String, dynamic> toJson() => _$UserToJson(this);
}

@JsonSerializable()
class Participant {
  final User user;
  final String role;
  final DateTime joinedAt;
  final String status;
  @JsonKey(name: '_id')
  final String id;

  Participant({
    required this.user,
    required this.role,
    required this.joinedAt,
    required this.status,
    required this.id,
  });

  factory Participant.fromJson(Map<String, dynamic> json) =>
      _$ParticipantFromJson(json);

  Map<String, dynamic> toJson() => _$ParticipantToJson(this);
}

@JsonSerializable()
class Checkpoint {
  final String? id;
  final String name;
  final String? description;
  final RunLocation location;
  final double radius;
  final int order;
  final List<CheckpointProgress> participantProgress;

  Checkpoint({
    this.id,
    required this.name,
    this.description,
    required this.location,
    required this.radius,
    required this.order,
    required this.participantProgress,
  });

  factory Checkpoint.fromJson(Map<String, dynamic> json) =>
      _$CheckpointFromJson(json);

  Map<String, dynamic> toJson() => _$CheckpointToJson(this);
}

@JsonSerializable()
class RunLocation {
  final double latitude;
  final double longitude;

  RunLocation({
    required this.latitude,
    required this.longitude,
  });

  factory RunLocation.fromJson(Map<String, dynamic> json) =>
      _$RunLocationFromJson(json);

  Map<String, dynamic> toJson() => _$RunLocationToJson(this);
}

@JsonSerializable()
class ChatMessage {
  final String? id;
  final String userId;
  final String message;
  final DateTime timestamp;

  ChatMessage({
    this.id,
    required this.userId,
    required this.message,
    required this.timestamp,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) =>
      _$ChatMessageFromJson(json);

  Map<String, dynamic> toJson() => _$ChatMessageToJson(this);
}

@JsonSerializable()
class CheckpointProgress {
  final String userId;
  final DateTime timestamp;
  final bool reached;

  CheckpointProgress({
    required this.userId,
    required this.timestamp,
    required this.reached,
  });

  factory CheckpointProgress.fromJson(Map<String, dynamic> json) =>
      _$CheckpointProgressFromJson(json);

  Map<String, dynamic> toJson() => _$CheckpointProgressToJson(this);
}

@JsonSerializable()
class RunMetrics {
  final String userId;
  final RunLocation? location;
  final double distance;
  final double? averagePace;
  final int? totalTime;

  RunMetrics({
    required this.userId,
    this.location,
    required this.distance,
    this.averagePace,
    this.totalTime,
  });

  factory RunMetrics.fromJson(Map<String, dynamic> json) =>
      _$RunMetricsFromJson(json);

  Map<String, dynamic> toJson() => _$RunMetricsToJson(this);
}
